### PR TITLE
executioner: move solver related params into FEProblem

### PR DESCRIPTION
**Pull request by rwcarlsen**
This is cleanup and preparation for more work on #5229.

### CODE REVIEW

**Comment by rwcarlsen** [comment]
Hints welcome if anybody knows at a glance what might be the issue.  Otherwise, I'll do some debugging tomorrow.

**Comment by friedmud** [comment]
Can you say a bit more about why you want to do this?  You're going to have to update every input file on the planet... and I'm not sure it gains us anything.

I'm not necessarily against it... but I think there needs to be some serious justification for this.  Every input file since the beginning of time has specified these parameters in the `Executioner` block...

Also: if we're going to move them out of `Executioner` then there needs to be a debate about _where_ to move them.  I don't necessarily think the `Problem` block is the right block either (most people don't even know that that block exists... and most input files don't define it).

**Comment by friedmud** [comment]
I see what you've done... you're trying to parse the parameters in the `Executioner` block but store them on the `Problem`?

I'm not for that change.

If the parameters don't belong with the `Executioner`... then let's move them.  If they do... then keep them there.

**Comment by friedmud** [comment]
Also: as a general rule: we need to move stuff OUT of `FEProblem`... not INTO it!  😄

**Comment by YaqiWang** [comment]
That is what I am about to say: FEProblem is too big...

**Comment by rwcarlsen** [comment]
Discussion with @permcody suggested that there was a disconnect between what "executioner" means in the input file and in the code.  I am much less acquainted with MOOSE history and can't strongly defend either side particularly well.  I agree that the much more elegant (for code) solution would be to maintain direct association between params in input file blocks and params in code.  In this chain:

```
Input-file ==> parameter spec/init ==> parameter usage
```

that association is broken between parameter spec/init and parameter usage.  This refactor doesn't fix the breakage, but moves the breakage to be between input file and parameter spec/init instead.  My opinion is that this is a less bad place to break the association.  Breaking every input file ever seems like not worth it, unless we are ready to make MOOSE 2.0.

This PR does not move those parameters from the Executioner block (I see you noticed that now).  Even though the solver parameters might not belong in FEProblem (which I definitely agree is too large), that is where they are now.  If someone comes along later and refactors the actual solver parameters out of the FEProblem class, they will probably want to have all the associated InputParameters with the new code (class) and actual member variables.  This preemptively decouples those InputParameters from the Executioner class where they were never stored/used already and makes them sit along-side the actual code they affect.

This also selfishly makes it easier for me to discover/see the actual executioner functionality free of things that were not super strongly related to the services it provides to the framework.

**Comment by friedmud** [comment]
If we're going to move them out... let's not move them to `FEProblem` directly.  And, I definitely understand not wanting to break every input file.

We've kind of been needing a `[Solver]` block for quite a while... but we've just continued sticking stuff in `Executioner` for now.

How about, as a start, we at least create a `SolverParams` (or whatever) object to stuff these parameters into.  For now, that object can hang off of `FEProblem` (or maybe even better: `NonlinearSystem` since that's where the solver is).

**Comment by rwcarlsen** [comment]
So FEProblem already has a ``SolverParams _solver_params`` member variable.  Currently it doesn't have much in it:

```c++
class SolverParams
{
public:
  SolverParams();

  Moose::SolveType _type;
  Moose::LineSearchType _line_search;
};
```

Would these options fit together well?  They currently live in several places already :/

```c++
  EquationSystems & esys = es();
  esys.parameters.set<Real> ("linear solver tolerance")= params.get<Real>("l_tol");
  esys.parameters.set<Real> ("linear solver absolute step tolerance") = params.get<Real>("l_abs_step_tol");
  esys.parameters.set<unsigned int> ("linear solver maximum iterations") = params.get<unsigned int>("l_max_its");
  esys.parameters.set<unsigned int> ("nonlinear solver maximum iterations") = params.get<unsigned int>("nl_max_its");
  esys.parameters.set<unsigned int> ("nonlinear solver maximum function evaluations") = params.get<unsigned int>("nl_max_funcs");
  esys.parameters.set<Real> ("nonlinear solver absolute residual tolerance") = params.get<Real>("nl_abs_tol");
  esys.parameters.set<Real> ("nonlinear solver relative residual tolerance") = params.get<Real>("nl_rel_tol");
  esys.parameters.set<Real> ("nonlinear solver absolute step tolerance") = params.get<Real>("nl_abs_step_tol");
  esys.parameters.set<Real> ("nonlinear solver relative step tolerance") = params.get<Real>("nl_rel_step_tol");

  getNonlinearSystem()._compute_initial_residual_before_preset_bcs = params.get<bool>("compute_initial_residual_before_preset_bcs");
  getNonlinearSystem()._l_abs_step_tol = params.get<Real>("l_abs_step_tol");
  getNonlinearSystem().setDecomposition(params.get<std::vector<std::string> >("splitting"));

  FileNameNoExtension restart_file_base(params.get<FileNameNoExtension>("restart_file_base"));
  if (!restart_file_base.empty())
    setRestartFile(restart_file_base);
```

Some in equation system, some in FEProblem's _app and _resurector members, some in nonlinear system member, etc.  Maybe just the EquationSystem params would fit in the SolverParams class?

**Comment by friedmud** [comment]
I think most of those would fit together rather well.

The `setDecomposition()` doesn't really belong there... I don't even know
what that's doing... it should be set as data on MooseMesh.

**Comment by fdkong** [comment]
> The `setDecomposition()` doesn't really belong there... I don't even know
> what that's doing... it should be set as data on MooseMesh.

`setDecomposition()` is used to set a flag to tell MOOSE we want to use the fieldSplit preconditioner and to do some preparation.

**Comment by friedmud** [comment]
Ah - I was thinking it was setting _mesh_ decomposition.  Maybe that method
needs to be a little more explanatory (`setFieldSplitType()` ?)

**Comment by fdkong** [comment]
> Ah - I was thinking it was setting _mesh_ decomposition. Maybe that method needs to be a little more explanatory (`setFieldSplitType()` ?)

`setFieldSplitTopLevel()`? The function will store where is the starting point of splitting.

**Comment by friedmud** [comment]
Sounds good to me.

File: framework/src/base/FEProblem.C [hunk 1]
```diff
@@ -98,6 +98,93 @@ InputParameters validParams<FEProblem>()
   return params;
 }
 
+//
+InputParameters
+FEProblem::executionerValidParams()
```
Comment (friedmud) [inline framework/src/base/FEProblem.C]: I understand what you're up to here... but having returned to this after a few months it still feels like `FEProblem` is not the right place for this stuff.

It really feels like we need a `SolverParams` object... and this function would go there.

Comment (friedmud) [inline framework/src/base/FEProblem.C]: * other than `restart_file_base`... that's not a "solver" parameter.

**Review by friedmud — CHANGES_REQUESTED** [review]
Good progress...

**Comment by rwcarlsen** [comment]
Agree this isn't a great solution.  We probably should hold on this until
we have a better idea about if/when/what we are going to do w.r.t.
executioner refactor.  That is why I haven't been actively working on this
anyway.

**Comment by friedmud** [comment]
Ok - should we close this for now then?

**Comment by fdkong** [comment]
> It really feels like we need a SolverParams object... and this function would go there.

@friedmud, we do have an object called `SolverParams` in the folder: framework/src/base.  The object has two members only, namely `solvetyper` and `linesearch`.  It is not a bad idea if we like to move some parameters there.

**Comment by rwcarlsen** [comment]
closed. created #8294 as reminder.

**Comment by friedmud** [comment]
Thanks @rwcarlsen 

@fdkong : Right, I agree that that's the right thing to do... but I think we also need a `Solver` block in the input file to go with it as well.

**Comment by YaqiWang** [comment]
Does solver here mean PETSc? I thought executioner is the solver of moose applications.

**Comment by friedmud** [comment]
Solver means "linear / nonlinear solver"

In my mind the Executioner is not a "solver"... but I know that in your case that's a bit more blurry.

**Pull request closed without merging**